src/daemon/server.py

```diff
@@ -0,0 +1,239 @@
+    process = services.get(service_name)
+    if process is None:
+        return 0
+    del services[service_name]
+    pid_path = pid_path_for_service(root_path, service_name)
+
+    log.info("sending term signal to %s", service_name)
+    process.terminate()
+    # on Windows, process.kill and process.terminate are the same,
+    # so no point in trying process.kill later
+    if process.kill != process.terminate:
+        count = 0
+        while count < delay_before_kill:
+            if process.poll() is not None:
+                break
+            await asyncio.sleep(1)
+            count += 1
+        else:
+            process.kill()
+            log.info("sending kill signal to %s", service_name)
+    r = process.wait()
+    log.info("process %s returned %d", process, r)
+    try:
+        pid_path_killed = pid_path.with_suffix(".pid-killed")
+        if pid_path_killed.exists():
+            pid_path_killed.unlink()
+        os.rename(pid_path, pid_path_killed)
+    except Exception:
+        pass
+
+    return 1
+
+
+def create_server_for_daemon(root_path):
+    routes = web.RouteTableDef()
+
+    services = dict()
+
+    @routes.get('/daemon/ping/')
+    async def ping(request):
+        return web.Response(text="pong")
+
+    @routes.get('/daemon/service/start/')
+    async def start_service(request):
+        service_name = request.query.get("service")
+
+        if not validate_service(service_name):
+            r = "unknown service"
+            return web.Response(text=str(r))
+
+        if service_name in services:
+            r = "already running"
+            return web.Response(text=str(r))
+
+        try:
+            process, pid_path = launch_service(root_path, service_name)
+            services[service_name] = process
+            r = "started"
+        except (subprocess.SubprocessError, IOError):
+            log.exception(f"problem starting {service_name}")
+            r = "start failed"
+
+        return web.Response(text=str(r))
+
+    @routes.get('/daemon/service/stop/')
+    async def stop_service(request):
+        service_name = request.query.get("service")
+        r = await kill_service(root_path, services, service_name)
+        return web.Response(text=str(r))
+
+    @routes.get('/daemon/service/is_running/')
+    async def is_running(request):
+        service_name = request.query.get("service")
+        process = services.get(service_name)
+        r = process is not None and process.poll() is None
+        return web.Response(text=str(r))
+
+    @routes.get('/daemon/exit/')
+    async def exit(request):
+
+        jobs = []
+        for k in services.keys():
+            jobs.append(kill_service(root_path, services, k))
+        if jobs:
+            done, pending = await asyncio.wait(jobs)
+        services.clear()
+
+        # TODO: fix this hack
+        asyncio.get_event_loop().call_later(5, lambda *args: sys.exit(0))
```

Comment:
Is this so that we can return the HTTP response?

Reply:
Yeah. Also, getting a clean exit can be tricky.